tests/test_parallel_validation.py

```diff
@@ -0,0 +1,125 @@
+    in_queue.put(RESET)
+    in_queue.put((20, create_tx.to_dict()))
+    in_queue.put((25, double_spend.to_dict()))
+    in_queue.put((30, transfer_tx.to_dict()))
+    in_queue.put(EXIT)
+
+    vw.run()
+
+    assert results_queue.get() == (0, create_tx)
+    assert results_queue.get() == (10, transfer_tx)
+    assert results_queue.get() == (20, False)
+    assert results_queue.get() == (0, create_tx)
+    assert results_queue.get() == (5, transfer_tx)
+    assert results_queue.get() == (20, create_tx)
+    assert results_queue.get() == (25, double_spend)
+    assert results_queue.get() == (30, False)
+
+
+def test_parallel_validator_routes_transactions_correctly(b, monkeypatch):
+    import os
+    from collections import defaultdict
+    import multiprocessing as mp
+    from json import dumps
+    from bigchaindb.parallel_validation import ParallelValidator
+
+    # We want to make sure that the load is distributed across all workers.
+    # Since introspection on an object running on a different process is
+    # difficult, we create an additional queue where every worker can emit its
+    # PID every time validation is called.
+    validation_called_by = mp.Queue()
+
+    # Validate is now a passthrough, and every time it is called it will emit
+    # the PID of its worker to the designated queue.
+    def validate(self, dict_transaction):
+        validation_called_by.put((os.getpid(), dict_transaction['id']))
+        return dict_transaction
+
+    monkeypatch.setattr(
+        'bigchaindb.parallel_validation.ValidationWorker.validate',
+        validate)
+
+    # Transaction routing uses the `id` of the transaction. This test strips
+    # down a transaction to just its `id`. We have two workers, so even ids
+    # will be processed by one worker, odd ids by the other.
+    transactions = [{'id': '0'}, {'id': '1'}, {'id': '2'}, {'id': '3'}]
+
+    pv = ParallelValidator(number_of_workers=2)
+    pv.start()
+
+    # ParallelValidator is instantiated once, and then used several times.
+    # Here we simulate this scenario by running it an arbitrary number of
+    # times.
+    # Note that the `ParallelValidator.result` call resets the object, and
+    # makes it ready to validate a new set of transactions.
+    for _ in range(2):
+        # First, we push the transactions to the parallel validator instance
+        for transaction in transactions:
+            pv.validate(dumps(transaction).encode('utf8'))
+
+        assert pv.result(timeout=1) == transactions
+
+        # Now we analize the transaction processed by the workers
+        worker_to_transactions = defaultdict(list)
+        for _ in transactions:
+            worker_pid, transaction_id = validation_called_by.get()
+            worker_to_transactions[worker_pid].append(transaction_id)
+
+        # The transactions are stored in two buckets.
+        for _, transaction_ids in worker_to_transactions.items():
+            assert len(transaction_ids) == 2
```

Comment:
We have deterministic assignments between `tx`s and workers. Can we strengthen this test to assert that worker0 gets `tx_0, 2` and worker1 gets `tx_1, 3`?

Reply:
https://github.com/bigchaindb/bigchaindb/pull/2489/commits/df824b07c1134891fb4802f84264f2aee02c04c2#diff-2772784c4ecd91053e0a2efcef70d208R125